Zend/zend_inheritance.c

```diff
@@ -2174,7 +2204,107 @@ static void zend_do_traits_method_binding(zend_class_entry *ce, zend_class_entry
+	 * of the constant from traits.
+	 */
+	size_t i;
+
+	if (colliding_ce == ce) {
+		for (i = 0; i < current_trait; i++) {
+			if (traits[i]
+				&& zend_hash_exists(&traits[i]->constants_table, constant_name)) {
+				return traits[i];
+			}
+		}
+	}
+	/* Traits don't have it, then the composing class (or trait) itself has it. */
+	return colliding_ce;
+}
+/* }}} */
+
+static bool do_trait_constant_check(zend_class_entry *ce, zend_class_constant *trait_constant, zend_string *name, zend_class_entry **traits, size_t current_trait) /* {{{ */
+{
+	uint32_t flags_mask = ZEND_ACC_PPP_MASK | ZEND_ACC_FINAL;
+
+	zval *zv = zend_hash_find_known_hash(&ce->constants_table, name);
+	if (zv == NULL) {
+		/* No existing constant of the same name, so this one can be added */
+		return true;
+	}
+
+	zend_class_constant *existing_constant = Z_PTR_P(zv);
+
+	if ((ZEND_CLASS_CONST_FLAGS(trait_constant) & flags_mask) != (ZEND_CLASS_CONST_FLAGS(existing_constant) & flags_mask) ||
+	    !check_trait_property_or_constant_value_compatibility(ce, &trait_constant->value, &existing_constant->value)) {
+		/* There is an existing constant of the same name, and it conflicts with the new one, so let's throw a fatal error */
+		zend_error_noreturn(E_COMPILE_ERROR,
+			"%s and %s define the same constant (%s) in the composition of %s. However, the definition differs and is considered incompatible. Class was composed",
+			ZSTR_VAL(find_first_constant_definition(ce, traits, current_trait, name, existing_constant->ce)->name),
+			ZSTR_VAL(trait_constant->ce->name),
+			ZSTR_VAL(name),
+			ZSTR_VAL(ce->name));
+	}
+
+	/* There is an existing constant which is compatible with the new one, so no need to add it */
+	return false;
+}
+/* }}} */
+
+static void zend_do_traits_constant_binding(zend_class_entry *ce, zend_class_entry **traits) /* {{{ */
+{
+	size_t i;
+
+	for (i = 0; i < ce->num_traits; i++) {
+		zend_string *constant_name;
+		zend_class_constant *constant;
+
+		if (!traits[i]) {
+			continue;
+		}
+
+		ZEND_HASH_MAP_FOREACH_STR_KEY_PTR(&traits[i]->constants_table, constant_name, constant) {
+			if (do_trait_constant_check(ce, constant, constant_name, traits, i)) {
+				zend_class_constant *ct = NULL;
+
+				ct = zend_arena_alloc(&CG(arena),sizeof(zend_class_constant));
+				memcpy(ct, constant, sizeof(zend_class_constant));
+				constant = ct;
+
+				if (Z_TYPE(constant->value) == IS_CONSTANT_AST) {
+					ce->ce_flags &= ~ZEND_ACC_CONSTANTS_UPDATED;
+					ce->ce_flags |= ZEND_ACC_HAS_AST_CONSTANTS;
+				}
+
+				/* Unlike interface implementations and class inheritances,
+				 * access control of the trait constants is done by the scope
+				 * of the composing class. So let's replace the ce here.
+				 * And the result of this, the values, doc comments, and
+				 * attributes associated with the copied zend_class_constant
+				 * are destroyed at the time of destroying the class entry of
+				 * the composing class.
+				 */
+				constant->ce = ce;
```

Comment:
So basically attributes on trait constants are "useless"? Would it be possible to check at compile time no attributes are used?

Reply:
> So basically attributes on trait constants are "useless"? 

Hmm? No, it's not. Maybe my wording is incorrect.

The sentences beginning with "And the result of this" are just to indicate the timing of the destruction of the data associated with the constant. Attributes on trait constants can be retrieved via either the trait itself or the composing class without problems.

For inheritance and interface implementation, the ce of zend_class_constant is kept as it was, while the same zend_class_constant is also inserted into the constants table of the inheriting child.

Destruction of class entries, such as [destroy_zend_class()](https://github.com/php/php-src/blob/822881b6fe88a271fe9d4c7f3b83083ff846fbbd/Zend/zend_opcode.c#L393), has a step to check if the ce of zend_class_constant in the constants table matches the class entry about to be destroyed. So the destruction timing of associated data, like attributes or values, is at the time of the destruction of the parent classes or interfaces that define the constants, not at the time of the destruction of the inheriting child. OTH, for trait constants, the destruction timing of each copied zend_class_constant is when each composing class is destroyed.

Reply:
Thinking again, I decided to remove the part that begins with "And the result of this".

It seems unlikely that this line will be overlooked when someone changes the implementation of access controls of trait constants. But it might be possible when changing the destruction of class entries.

On PR comments, it may now help people read the implementation, but if embedded into the code, it may cause unnecessary misunderstandings in the future.

Reply:
Removed that part and also added test cases for attributes and doc comments.

Reply:
I think I got confused because you are talking about destroying the content, when in fact it was already copied into the class the trait is used.